src/librustc_typeck/collect/type_of.rs

```diff
@@ -17,29 +17,141 @@ use rustc_trait_selection::traits;
 use super::ItemCtxt;
 use super::{bad_placeholder_type, is_suggestable_infer_ty};
 
+/// Computes the relevant generic parameter for a potential generic const argument.
+///
+/// This should be called using the query `tcx.opt_const_param_of`.
+pub(super) fn opt_const_param_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<DefId> {
+    use hir::*;
+
+    let hir_id = tcx.hir().as_local_hir_id(def_id);
+
+    if let Node::AnonConst(_) = tcx.hir().get(hir_id) {
```

Comment:
I think you can make this cheaper incremental-wise by checking `tcx.def_kind(def_id)` first, because that won't add a dependency to the entire definition pointed to by `def_id`, just the simple `DefKind` enum (but feel free to file this under further experimentation).